generator/internal/discovery_file.cc

```diff
@@ -0,0 +1,116 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "generator/internal/discovery_file.h"
+#include "generator/internal/codegen_utils.h"
+#include "google/cloud/internal/absl_str_join_quiet.h"
+#include "absl/strings/str_format.h"
+#include <google/protobuf/io/printer.h>
+#include <google/protobuf/io/zero_copy_stream_impl.h>
+#include <fstream>
+#ifdef _WIN32
+#include <direct.h>
+#else
+#include <sys/stat.h>
+#endif  // _WIN32
+
+namespace google {
+namespace cloud {
+namespace generator_internal {
+namespace {
+
+void MakeDirectory(std::string const& path) {
```

Comment:
We also have one of these in:

https://github.com/googleapis/google-cloud-cpp/blob/0633e0deda8d91eda85ee12a597e08f3d01efd24/generator/internal/scaffold_generator.cc#L144-L150

Maybe it is time to refactor them to a [`create_directory()`](https://en.cppreference.com/w/cpp/filesystem/create_directory) in 

https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/internal/filesystem.h
https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/internal/filesystem.cc

Reply:
Whichever way you go, 777 is usually the right choice for the mode (unless there are special security concerns) so that the umask has a chance to work.